### PR TITLE
Add optimized `PixtralImageProcessorFast`

### PR DESCRIPTION
**Pull request by mgoin**
# What does this PR do?

This PR implements a fast image processor for Pixtral. Follows issue https://github.com/huggingface/transformers/issues/33810.

The key acceleration comes from replacing Pillow/Numpy tensors and functions (resize, rescale, normalize) with torch tensors and torchvisionv2 functions. It comes along with support for `torch.compile` and passing `device="cuda"` during inference to process the input on GPU. One limitation is that only `return_tensors="pt"` will be supported. 

Usage
```python
from transformers import AutoImageProcessor

slow_processor = AutoImageProcessor.from_pretrained("mistral-community/pixtral-12b", use_fast=False)
fast_processor = AutoImageProcessor.from_pretrained("mistral-community/pixtral-12b", use_fast=True)
compiled_processor = torch.compile(fast_processor, mode="reduce-overhead")
```

From simple benchmarking with a single image of size `[3, 876, 1300]`, I see 6x to 10x speedup

image

```
--------------------------------------------------
Slow Processor (PIL Image) Statistics (milliseconds):
          Mean: 23.680
        Median: 23.098
       Std Dev: 2.240
           Min: 21.824
           Max: 36.064

--------------------------------------------------
Fast Processor (PIL Image) Statistics (milliseconds):
          Mean: 3.759
        Median: 3.762
       Std Dev: 0.133
           Min: 3.556
           Max: 4.223

--------------------------------------------------
Compiled Processor (PIL Image) Statistics (milliseconds):
          Mean: 4.632
        Median: 4.794
       Std Dev: 1.086
           Min: 3.488
           Max: 11.707

--------------------------------------------------
Slow Processor (Torch Image) Statistics (milliseconds):
          Mean: 22.331
        Median: 21.878
       Std Dev: 1.821
           Min: 21.316
           Max: 36.603

--------------------------------------------------
Fast Processor (Torch Image) Statistics (milliseconds):
          Mean: 2.242
        Median: 2.209
       Std Dev: 0.164
           Min: 2.182
           Max: 3.803

--------------------------------------------------
Compiled Processor (Torch Image) Statistics (milliseconds):
          Mean: 2.125
        Median: 2.117
       Std Dev: 0.073
           Min: 2.062
           Max: 2.594
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qubvel** [comment]
Hi @mgoin! Sounds great! Thanks for working on this 🤗 

cc @yonigozlan maybe if you have bandwidth

**Review by yonigozlan — COMMENTED** [review]
Thanks a lot @mgoin for working on this! Looks great to me, just mentioned some minor things to fix.

To be transparent, the current plan for fast image processors is to make a strong BaseImageProcessorFast and several image processing Mixins, in order to make adding new fast image processor much simpler. All that to say that this processor might change again in the future. Meanwhile, I think it would be great to have as is, because as you can see it makes a huge performance difference, and this fast image processor in particular doesn't require a huge diff (compared to the DETR ones for example).

Also, this would be the first fast image processor used in a processor, and I don't think there is a mechanism to use it with AutoProcessor yet (I might be wrong), so this is also something that will need to be added soon. 
But I'm curious then to know how you are using it with a PixtralProcessor, I'm guessing you manually instantiate it with:
```python
from transformers import PixtralProcessor, AutoImageProcessor, AutoTokenizer

fast_image_processor = AutoImageProcessor.from_pretrained("mistral-community/pixtral-12b", use_fast=True)
tokenizer = AutoTokenizer.from_pretrained("mistral-community/pixtral-12b")

processor = PixtralProcessor(fast_image_processor, tokenizer)
```
Is that correct?

Thanks again!

**Comment by mgoin** [comment]
Thanks for the review and context @yonigozlan ! I will look into it later today. Yes you are correct about using it within a processor. I have tested this works within vLLM simply by adding `use_fast=True` to our `AutoProcessor.from_pretrained()` call [here](https://github.com/vllm-project/vllm/blob/da7e702c6fae521bf8633affb8fe7b834f5cb94b/vllm/transformers_utils/processor.py#L18-L22).

One bug I noticed is that if I specify `use_fast=True` and there isn't a Fast version of the ImageProcessor available, I get an exception. I can look into this, but would be good to get clarity that this is unintended behavior.

**Comment by yonigozlan** [comment]
Oh great news that it already works with AutoProcessor. As I said this is the first fast image processor used in a processor so it was not guaranteed :).

> One bug I noticed is that if I specify use_fast=True and there isn't a Fast version of the ImageProcessor available, I get an exception. I can look into this, but would be good to get clarity that this is unintended behavior.

Yes this is the same right now when using ImageProcessingAuto. I don't think it should be that way though, especially as more and more people will want to use fast image processors by default. I'll open a PR to fix this.

Current plan is:
- keep use_fast to False by default in Auto classes during a deprecation cycle, fall back on slow image processor (with a warning) if use_fast is set to True and no fast image processor exists
- set use_fast to True by default in Auto classes after the deprecation cycle (by then most models will hopefully have a fast image processor), still fall back on slow if no fast image processor exists.

The deprecation cycle is needed as there are slight differences in outputs when using torchvision vs PIL, see this PR https://github.com/huggingface/transformers/pull/34785 for more info.

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+    is_torch_available,
+    is_torchvision_available,
+    is_torchvision_v2_available,
+    is_vision_available,
+    logging,
+)
+from .image_processing_pixtral import (
+    BatchMixFeature,
+    convert_to_rgb,
+    get_resize_output_image_size,
```
[the remark this answers is not in the document]
Comment (mgoin) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: - `BatchMixFeature` is essentially torch-only
- `convert_to_rgb` only does something if the image is a `PIL.Image.Image`, otherwise it just returns the image
- `get_resize_output_image_size` just uses `image.shape` so it doesn't make a difference for either numpy or torch

So I don't see where an issue would arise. I can update the type of the argument to make it clear

[the remark this answers is not in the document]
Comment (yonigozlan) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: Sorry I meant only `get_resize_output_image_size`. `np.ceil` is also used there.
But agreed there are no issues with `BatchMixFeature ` and `convert_to_rgb`

**Comment by ArthurZucker** [comment]
Feel free to ping us for another round of review! 🚀

File: src/transformers/image_utils.py [hunk 2]
```diff
@@ -447,6 +448,44 @@ def validate_preprocess_arguments(
...
+        resample=resample,
+    )
+    # Extra checks for ImageProcessorFast
+    if return_tensors != "pt":
+        raise ValueError("Only returning PyTorch tensors is currently supported.")
+
+    if data_format != ChannelDimension.FIRST:
+        raise ValueError("Only channel first data format is currently supported.")
+
+
```
Comment (yonigozlan) [inline src/transformers/image_utils.py:488]: Thanks for that 🤗

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
```
Comment (yonigozlan) [inline src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: I still think it would be better to rewrite this function with `torch.ceil` instead of `np.ceil`

**Review by yonigozlan — COMMENTED** [review]
Thanks for iterating!
LGTM after adding a `get_resize_output_image_size` with torch instead of numpy

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgoin) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: Ah I didn't notice the np.ceil, thanks for pointing out

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py (hunk 3 above)
Comment (mgoin) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: Sorry I missed your comment response, putting this in now

**Comment by mgoin** [comment]
Hi @yonigozlan what do you think about the use of `math.ceil`?

**Comment by yonigozlan** [comment]
> Hi @yonigozlan what do you think about the use of math.ceil?

Sorry to be annoying with this but I'd prefer `torch.ceil`. I'd say we leave the `get_resize_output_image_size` function as it is (with np.ceil) in `image_processing_pixtral`, and rewrite one with torch in `image_processing_pixtral_fast`.

**Comment by mgoin** [comment]
Sorry if I am misunderstanding something @yonigozlan  but `torch.ceil` only works on [`torch.Tensor`](https://pytorch.org/docs/stable/generated/torch.ceil.html). Since we are working with `height`/`width` as `int` primitives I cannot substitute `torch.ceil`. 

Are you suggesting that I wrap the values in a Tensor just to extract them back out as primitives? This is inefficient and unneeded to work with `torch.compile`

```python
height = int(np.ceil(height / ratio))
# versus
height = int(torch.ceil(torch.tensor(height / ratio)).item())
```

Considering [`torchvision.transforms.v2.functional.resize`](https://pytorch.org/vision/main/generated/torchvision.transforms.v2.functional.resize.html) takes in size as a list of ints (`size: Optional[List[int]]`), I don't see the benefit of using torch.Tensor for the sizes

**Comment by yonigozlan** [comment]
Hi @mgoin,
Sorry about that, I went over the code too quickly and completely missed that we were dealing with floats and not arrays in the np.ceil. So what you did with `ImageInput` and `math.ceil` in `image_processing_pixtral.get_resize_output_image_size` makes perfect sense, you can ignore my last few messages🤗. 
LGTM then! :)

File: src/transformers/models/pixtral/image_processing_pixtral_fast.py (hunk 3 above)
Comment (yonigozlan) [reply src/transformers/models/pixtral/image_processing_pixtral_fast.py:43]: Just saw this, indeed my mistake :)

**Comment by mgoin** [comment]
Nice thanks for clarification! Hi @ArthurZucker would you mind signing off?

**Comment by ArthurZucker** [comment]
Yep having a look!

**Review by ArthurZucker — APPROVED** [review]
LGTM, sorry for the delay! My main question is for @yonigozlan , do we have compile tests for image processor fast?

**Comment by ArthurZucker** [comment]
Thanks a lot @mgoin for this contribution! 🔥

**Comment by yonigozlan** [comment]
> LGTM, sorry for the delay! My main question is for @yonigozlan , do we have compile tests for image processor fast?
I don't think so, I'll open a PR for that!

**Comment by ArthurZucker** [comment]
Cool thanks!
